# scatac_root

The regulatory landscape of Arabidopsis thaliana roots at single-cell resolution

Authors: Michael W. Dorrity, Cristina M. Alexandre, Morgan O. Hamm, Anna-Lena Vigil, Stanley Fields, Christine Queitsch and Josh Cuperus

Abstract: The scarcity of accessible sites that are dynamic or cell type-specific in plants may be due in part to tissue heterogeneity in bulk studies. To assess the effects of tissue heterogeneity, we apply single-cell ATAC-seq to Arabidopsis thaliana roots and identify thousands of differentially accessible sites, sufficient to resolve all major cell types of the root. We find that the entirety of a cell's regulatory landscape and its transcriptome independently capture cell type identity. We leverage this shared information on cell identity to integrate accessibility and transcriptome data to characterize developmental progression, endoreduplication and cell division. We further use the combined data to characterize cell type-specific motif enrichments of transcription factor families and link the expression of family members to changing accessibility at specific loci, resolving direct and indirect effects that shape expression. Our approach provides an analytical framework to infer the gene regulatory networks that execute plant development.


This markdown contains useful code blocks for producing the main results of the manuscript and exploring further the relationships between scATAC-seq and scRNA-seq data in the Arabidopsis root. 